Spy support

dreef3 opened the pull request:
Hi,

This is an attempt to reuse the existing `Mocker` class to provide a functionality for spying on the existing object.

Closes #33

NagRock commented:
@dreef3 thanks for another pull request! I will check it and give feedback.

dreef3 commented:
@NagRock this one definitely needs more tests, will add them if the approach to implementation looks good.

NagRock commented on test/spy.spec.ts:
```diff
@@ -0,0 +1,81 @@
+
+    describe("calling a mocked method", () => {
+        it("delegates the call to a mock", () => {
+            let foo = new Real();
```
please use given/when/then convention

NagRock commented on README.md:
````diff
@@ -306,6 +307,18 @@ const foo: SampleGeneric<SampleInterface> = instance(mockedFoo);
+
+``` typescript
+const foo: Foo = new Foo();
+const mockedFoo = spy(foo);
````
I think that `spiedFoo` would be better name. What do you think?

NagRock commented on test/spy.spec.ts:
```diff
@@ -0,0 +1,81 @@
+
+            foo.foo(42);
+
+            expect(capture(mocked.foo).last()).toEqual([42]);
```
Why we need to use array `[42]` here instead of number `42`?

dreef3 replied:
Because `ArgCaptor.last()` returns an array of arguments

NagRock commented on test/spy.spec.ts:
```diff
@@ -0,0 +1,81 @@
+    });
+
+    describe("resetting", () => {
+        it("restores the call to a real method", () => {
```
Nice test! 👍

NagRock commented:
Nice that `calledBefore` and `calledAfter` functionality still works!

I need some more time to check API, because when using mock we got to use `instance` while using `spy` we're not. Please give me one more day to check it :) And one more time thanks for contributing!

NagRock commented:
@dreef3 hi again, it looks great! Two things. We need some more tests (for example call order testing `calledBefore`, `calledAfter` and call count `once()`, `twice()` etc.)

And second thing:

```
            let foo = new Real();
            let mocked = spy(foo);

            when(mocked.foo(1)).thenReturn(3);

            expect(foo.foo(1)).toBe(3);
            expect(foo.foo(2)).toBe(2); // here should be called real implementation but we get null
```

If you need some more information or help with something then ping me!

viman commented on src/Mock.ts:
```diff
@@ -59,36 +58,46 @@ export class Mocker {
+                }
+
+                prototype = prototype.__proto__;
+                this.createMethodStubsFromOwnPropertyNames(prototype);
```
See the comment above

viman commented on src/Mock.ts:
```diff
@@ -155,27 +164,34 @@ export class Mocker {
             prototype = prototype.__proto__;
             if (prototype && prototype !== Object.prototype) {
-                this.createInstanceActionListenersFromPrototypeOwnPropertyDescriptors(prototype);
+                this.createInstanceActionListenersFromOwnPropertyDescriptors(prototype);
```
If you are changing recursion into while loop please be consistent and implement it here also.

dreef3 replied:
Reverted the while loop, only needed to check for Object.prototype at the beginning to catch the case with plain object without inheritance hierarchy.

viman commented on src/Mock.ts:
```diff
@@ -185,16 +201,21 @@ export class Mocker {
             prototype = prototype.__proto__;
             if (prototype && prototype !== Object.prototype) {
-                this.createInstanceActionListenersFromPrototypeOwnPropertyNames(prototype);
+                this.createInstanceActionListenersFromOwnPropertyNames(prototype);
```
See the comment above.

viman commented on src/spy/RealMethod.ts:
```diff
@@ -0,0 +1,4 @@
+export class RealMethod {
+    constructor(public descriptor: PropertyDescriptor,
+                public instance: boolean) {}
+}
```
Please add a new line on the end of files.

dreef3 replied:
done

viman commented:
Great job @dreef3 with the spies! 👍 Please see my comments.

dreef3 commented on src/Mock.ts:
```diff
@@ -59,36 +58,46 @@ export class Mocker {
+                }
+
+                prototype = prototype.__proto__;
+                this.createMethodStubsFromOwnPropertyDescriptors(prototype);
```
Reverted to recursion, merged two methods into one

dreef3 commented:
@NagRock,

1. I've added some more tests for verification, not sure it's necessary to test all verification methods because it would duplicate existing `verification.spec.ts`.
2. Fixed this case

NagRock commented:
@dreef3 Thanks for changes! You right that it would duplicate existing specs so don't do this. I will try to parameterise it later.

NagRock commented:
@dreef3 merged! Once again thanks for this feature!
